src/newton_ac_powerflow.jl

```diff
@@ -0,0 +1,726 @@
+    V .= Vm .* exp.(1im .* Va)
+
+    # early return if only ref buses present - no need to solve the power flow
+    converged = npvpq == 0
+    if converged
+        # if only ref buses present, we do not need to enter the power flow loop
+        x = _calc_x(data, V, Va, Vm, Ybus, n_buses)
+        return (converged, x)
+    end
+
+    # we need to define lookups for mappings of pv, pq buses onto the internal J indexing
+    pvpq_lookup = zeros(Int64, maximum([ref; pvpq]) + 1)
+    pvpq_lookup[pvpq] .= 1:npvpq
+    pq_lookup = zeros(Int64, maximum([ref; pvpq]) + 1)
+    pq_lookup[pq] .= 1:npq
+
+    # define the internal J indexing using the lookup arrays
+    j_pvpq = pvpq_lookup[pvpq]
+    j_pq = npvpq .+ pq_lookup[pq]
+
+    # indices for updating of V
+    dx_Va_pv = Vector{Int64}([1:npv...])
+    dx_Va_pq = Vector{Int64}([(npv + 1):(npv + npq)...])
+    dx_Vm_pq = Vector{Int64}([(npv + npq + 1):(npv + 2 * npq)...])
+
+    Sbus =
+        data.bus_activepower_injection[:, time_step] -
+        data.bus_activepower_withdrawals[:, time_step] +
+        1im * (
+            data.bus_reactivepower_injection[:, time_step] -
+            data.bus_reactivepower_withdrawals[:, time_step]
+        )
+
+    # Pre-allocate mis and F and create views for the respective real and imaginary sections of the arrays:
+    mis = zeros(Complex{Float64}, length(V))
+    F = zeros(Float64, npvpq + npq)
+    mis .= V .* conj.(Ybus * V) .- Sbus
+    F .= [real(mis[pvpq]); imag(mis[pq])]
+
+    # preallocate Jacobian matrix and arrays for calculating dSbus_dVa, dSbus_dVm
+    rows, cols = SparseArrays.findnz(Ybus)
+
+    #diagV = sparse(1:n_buses, 1:n_buses, V)
+    diagV = LinearAlgebra.Diagonal(V)
+    diagIbus_diag = zeros(Complex{Float64}, size(V, 1))
+    diagIbus = LinearAlgebra.Diagonal(diagIbus_diag)
+
+    diagVnorm = LinearAlgebra.Diagonal(V ./ abs.(V))
+
+    Ybus_diagVnorm = sparse(rows, cols, Complex{Float64}(0))
+    conj_Ybus_diagVnorm = sparse(rows, cols, Complex{Float64}(0))
+    diagV_conj_Ybus_diagVnorm = sparse(rows, cols, Complex{Float64}(0))
+    conj_diagIbus = conj.(diagIbus)
+    conj_diagIbus_diagVnorm = conj.(diagIbus)
+    Ybus_diagV = sparse(rows, cols, Complex{Float64}(0))
+    conj_Ybus_diagV = sparse(rows, cols, Complex{Float64}(0))
+
+    dSbus_dVm = sparse(rows, cols, Complex{Float64}(0))
+    dSbus_dVa = sparse(rows, cols, Complex{Float64}(0))
+    r_dSbus_dVa = sparse(rows, cols, Float64(0))
+    r_dSbus_dVm = sparse(rows, cols, Float64(0))
+    i_dSbus_dVa = sparse(rows, cols, Float64(0))
+    i_dSbus_dVm = sparse(rows, cols, Float64(0))
+
+    # maybe use this in the future?
+    # pvpq_rows = pvpq_lookup[rows][pvpq_lookup[rows] .!= 0]
+    # pvpq_cols = pvpq_lookup[cols][pvpq_lookup[cols] .!= 0]
+    # pq_rows = pq_lookup[rows][pq_lookup[rows] .!= 0] 
+    # pq_cols = pq_lookup[cols][pq_lookup[cols] .!= 0]
+
+    J_block = sparse(rows, cols, Float64(0), maximum(rows), maximum(cols))
```

Comment:
Wrap this in a function like `get_jacobian` so that we can improve the return of this later